Source/WTF/wtf/ThreadSafeWeakHashSet.h

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <wtf/Algorithms.h>
+#include <wtf/HashSet.h>
+#include <wtf/ThreadSafeWeakPtr.h>
+
+namespace WTF {
+
+template<typename T>
+class ThreadSafeWeakHashSet final {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+
+    class const_iterator {
+    public:
+        using iterator_category = std::forward_iterator_tag;
+        using value_type = T;
+        using pointer = const value_type*;
+        using reference = const value_type&;
+
+        T* get() const { return m_strongReferences[m_position].ptr(); }
+        T& operator*() const { return *get(); }
+        T* operator->() const { return get(); }
+
+        const_iterator& operator++()
+        {
+            m_position++;
+            return *this;
+        }
+
+        bool operator==(const const_iterator& other) const
+        {
+            // Not a complete comparison, but only used to compare an iterator against the end.
+            ASSERT_UNUSED(other, !other.m_strongReferences.size());
+            return m_position >= m_strongReferences.size();
```

Comment:
Why can't we `return m_position == other.m_position;` ? We'd probably use `std::numeric_limits<size_t>::max()` for the position of the `end` iterator. I think we'd also need to set m_position to `std::numeric_limits<size_t>::max()` in operator++ in the case where `m_position` is part the end of `m_strongReferences`.

Alternatively, we could not have an iterator at all and require all call sites to use forEach(), it is unclear there is really a need to expose an iterator, especially if this hash set can get modified from another thread. It wouldn't "look" safe even if you wrote your iterator so that it is.

Reply:
I like the idea of removing the iterator completely.